src/DSP/Reverter.cpp

```diff
@@ -0,0 +1,282 @@
+        case NOTEONOFF:
+            if ( (reverse_index >= recorded_samples && state == PLAYING) || // finished playing OR
+                 (reverse_index >= max_delay_samples && state == PLAYING) ||// note is too long buffer ends here OR
+                 (mean_abs_value < 0.001f && state == RECORDING) )          // note has decayed
+                handleStateChange();
+            break;
+        default:
+            {}
+            break;
+    }
+}
+
+void Reverter::handleStateChange() {
+    if (state == RECORDING) {
+        recorded_samples = reverse_index;
+        state = PLAYING;
+    } else if (state == PLAYING) {
+        state = IDLE;
+    }
+    switchBuffers();
+}
+
+void Reverter::sync(float pos) {
+    if (state == IDLE) {
+        reset();
+        state = RECORDING;
+        reverse_index = 0;
+    } else {
+        syncPos = pos + reverse_index;
+        doSync = true;
+    }
+}
+
+void Reverter::updateReaderPosition(int i) {
+    //                 buffersize
+    //                  |<---->|
+    // ---+------+------+------+------+------+---
+    //    |      |      |      |      |      |
+    // ...|  b4  |  b5  |  b6  |  b7  |  b8  |...
+    // ---+------+------+------+------+------+---
+    //             ^           ^             ^
+    //             |           L pos_start   L pos_writer
+    //             L pos_reader
+    //             <--><------->
+    //              |       L reverse_index
+    //              L phase_offset
+    // //////////////////////////////////////////////////
+
+    // update reading position
+    pos_reader = fmodf(float(pos_start + mem_size) - (reverse_index + phase_offset), mem_size);
+}
+
+void Reverter::crossfadeSamples(float *smp, int i) {
+    if (fade_counter < fading_samples) {
+        float fadeinFactor = static_cast<float>(fade_counter) / static_cast<float>(fading_samples);
+        float fadeoutFactor = 1.0f - fadeinFactor;
+        fade_counter++;
+
+        if (state == IDLE) {
+            // in IDLE only the fade out part is needed
+            smp[i] = fadeoutFactor * sampleLerp(input, fmodf(pos_reader + mem_size + delta_crossfade, mem_size));
+        } else {
+            smp[i] = applyFade(fadeinFactor, fadeoutFactor);
+        }
+    } else {
+        smp[i] = (state == PLAYING) ? sampleLerp(input, pos_reader) : 0.0f;
+    }
+}
+
+float Reverter::applyFade(float fadein, float fadeout) {
+    if (syncMode == NOTEON || syncMode == NOTEONOFF) {
+        return fadein * sampleLerp(input, pos_reader);
+    } else {
+        return fadein * sampleLerp(input, pos_reader) + fadeout * sampleLerp(input, fmodf(pos_reader + mem_size + delta_crossfade, mem_size));
+    }
+}
+
+void Reverter::setdelay(float value) {
+
+    if (value == delay)
+        return;
+    else
+        delay = value;
+
+    // update mem_size since it depends on delay
+    setcrossfade(crossfade);
+    // recalculate global offset using new delay
+    global_offset = fmodf(tRef, delay);
```

Comment:
For a delay of 0, in NoteOnOff, I get no reversed sound at all, or at least, it sounds weird. Maybe this happens with `delay` being 0 here, and `fmodf` with 0 as second parameter is illegal?

Reply:
there is 1 added to Pdelay befor using it.